keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+Through a feature flag the new controller implementation is controlled.
+- alpha (disabled, can be enabled)
+- beta (enabled, can be disabled)
+- GA (enabled, cannot be disabled)
+
+API will track the controller since GA-ing the API implies GA-ing the scalable implementation of the API.
+
+### Existing controller
+
+The current implementation of the CronJob controller is different that the other workload controllers. GA workload controllers use informers and caches to reduce the load on API server. Whereas the cronjob controller does a periodic poll and sweep of all the objects and acts on them. The CronJob controller has only one worker doing this.
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L98).
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L136)
+
+This is not a scalable design and ends up loading the API server. Also this does not follow the [recommended guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md) for building controllers.
+
+### New controller
+
+With this approach we aim to:
+1. Reduce the potential scale issues when using lots of CronJob objects
+2. Reduce load on API server in such cases.
+3. Reduce memory usage
+
+#### Informers and Caches
+
+To reduce the need to list all Jobs and CronJobs frequently to reconcile, we propose to use Informers and WorkQueue based architecture. We would be sharing the [same informer cache](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L450) as the Job controller uses.
+
+We will follow a controller structure similar to existing workload controllers and as [outlined in the guideline](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure).
+```golang
+
+// CronJobControllerV2 is 2nd generation controller for cronjobs.
+// It is using informers and workqueue to improve its performance over
+// the old controller.
+type CronJobControllerV2 struct {
+	queue	   			workqueue.RateLimitingInterface
+	recorder   			record.EventRecorder
+
+	jobControl 			jobControlInterface
+	cronJobControl  		cronJobControlInterface
+
+	jobLister  			batchv1listers.JobLister
+	cronJobLister 			CronJobLister
+
+	jobListerSynced 		cache.InformerSynced
+	cronJobListerSynced		cache.InformerSynced
+}
+
+// CronJobControllerConfiguration contains controller config
+type CronJobControllerConfiguration struct {
+	// concurrentCronJobSyncs is the number of cronjob objects that are
+	// allowed to sync concurrently. Larger number = more responsive
+	// but more CPU (and network) load.
+	ConcurrentCronJobSyncs int32
+}
+
+
+// Controller code
```

Comment:
nit: I'm not entirely sure it makes sense to put this code here (it's pretty much the same as in all other controllers) and it adds unnecessary 40 lines imho.

Reply:
Removed.